File: src/utils/ConstantHeadingDifferentialEquation.ts
import type { Point2D, CubicBezierCurve } from '../types/geometry';

export class ConstantHeadingDifferentialEquation {
  private theta: number;
  private vMax: number;
  private mass: number;
  private muK: number;
  private c1: number;
  private path: CubicBezierCurve;

  constructor(
    theta: number,
    vMax: number,
    mass: number,
    muK: number,
    c1: number,
    path: CubicBezierCurve
  ) {
    this.theta = theta;
    this.vMax = vMax;
    this.mass = mass;
    this.muK = muK;
    this.c1 = c1;
    this.path = path;
  }

  getDimension(): number {
    return 2;
  }

  computeDerivatives(v: number, state: number[], derivatives: number[]): void {
    const point: Point2D = { x: state[0], y: state[1] };
    const t = this.path.pathInversion(point);
    const pathPrime = this.path.computePathPrimes(t);

    // Create vector operations
    const pathPrimeNorm = Math.sqrt(pathPrime.x * pathPrime.x + pathPrime.y * pathPrime.y);
    const dotProduct = 
      pathPrime.x * Math.cos(this.theta) + 
      pathPrime.y * Math.sin(this.theta);

    const pathAngle = Math.atan2(pathPrime.y, pathPrime.x);
    const scalar = (
      dotProduct * this.vMax - 
      pathPrimeNorm * (
        (this.muK * this.mass) + 
        this.c1 * Math.abs(Math.sin(this.theta - pathAngle))
      )
    ) / (pathPrimeNorm * pathPrimeNorm);

    derivatives[0] = scalar * pathPrime.x;
    derivatives[1] = scalar * pathPrime.y;
  }
} 

File: src/fpa/java/ConstantHeadingInterpolation/ConstantHeadingDifferentialEquation.java
package java.ConstantHeadingInterpolation;

import org.apache.commons.math3.exception.DimensionMismatchException;
import org.apache.commons.math3.exception.MaxCountExceededException;
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;
import org.apache.commons.math3.ode.FirstOrderDifferentialEquations;
import org.apache.commons.math3.util.FastMath;

import java.MathUtil.CubicBezierCurve;
import java.awt.geom.Point2D;

public class ConstantHeadingDifferentialEquation implements FirstOrderDifferentialEquations {
    private final double theta;
    private final double v_max; //inches per second
    private final double mass; //kg
    private final double mu_k;
    private final double c_1;
    private final CubicBezierCurve path;

    public ConstantHeadingDifferentialEquation(double theta, double vMax, double mass, double muK, double c1, CubicBezierCurve path) {
        this.theta = theta;
        v_max = vMax;
        this.mass = mass;
        mu_k = muK;
        c_1 = c1;
        this.path = path;
    }

    @Override
    public int getDimension() {
        return 2;
    }

    @Override
    public void computeDerivatives(double v, double[] doubles, double[] doubles1) throws MaxCountExceededException, DimensionMismatchException {
        double t = path.pathInversion(new Point2D.Double(doubles[0], doubles[1]));
        Point2D.Double pathPrime = path.computePathPrimes(t);
        Vector2D pathPrimes = new Vector2D(pathPrime.getX(), pathPrime.getY());

        double scalar = (pathPrimes.dotProduct(new Vector2D(FastMath.cos(theta), FastMath.sin(theta))) * v_max - pathPrimes.getNorm() * ((mu_k * mass) + c_1 * FastMath.abs(FastMath.sin(theta-FastMath.atan2(pathPrime.getY(), pathPrime.getX())))))/pathPrimes.getNormSq();

        doubles1[0] = scalar * pathPrimes.getX();
        doubles1[1] = scalar * pathPrimes.getY();
    }
}
